File: typescript/src/client/whm.ts
import { request, CpanelApiError } from "./base.js";

export interface WhmResponse<T = unknown> {
  metadata: {
    version: number;
    result: number;
    reason: string;
    command: string;
  };
  data: T;
}

export async function whm<T = unknown>(
  func: string,
  params?: Record<string, string>,
): Promise<T> {
  const allParams = { "api.version": "1", ...params };
  const raw = await request<WhmResponse<T>>(
    "whm",
    `/json-api/${func}`,
    allParams,
  );

  if (raw.metadata.result !== 1) {
    throw new CpanelApiError(0, raw.metadata.reason || "Unknown WHM error", raw);
  }

  return raw.data;
}

File: python/src/cpanel_mcp/client/whm.py
"""WHM API 1 client (port 2087, admin-level operations)."""

from ..config import CpanelConfig
from .base import CpanelApiError, request


async def whm_call(
    config: CpanelConfig,
    function: str,
    params: dict[str, str] | None = None,
) -> dict:
    all_params = {"api.version": "1", **(params or {})}
    raw = await request(config, "whm", f"/json-api/{function}", all_params)
    metadata = raw.get("metadata", {})
    if metadata.get("result") != 1:
        raise CpanelApiError(0, metadata.get("reason", "Unknown WHM error"), raw)
    return raw.get("data", {})
